File: Week3_Intro_HashTables/single-number.py
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        """
        Finds the number that only occurs once
        in nums, where every other element 
        occurs twice.
        """
        
        # Make frequency table
        freq_table = {}
        
        # Compute frequency. Very important syntax in dictionaries
        for i in nums:
            if i in freq_table:
                freq_table[i] += 1
            else:
                freq_table[i] = 1
        
        # Find number that only occurs once.
        for num, freq in freq_table.items():
            if freq == 1:
                return num
        

File: Week3_Intro_HashTables/single-number.java
import java.util.HashMap;

class Solution {
    public int singleNumber(int[] nums) {
        /* Create Frequency Table */
        HashMap<Integer, Integer> freq_table = new HashMap<>();

        /*
         * Loop through Array. If value does not exist in HashMap, 
         * then set default to 1. If not, add 1 to existing value
         */
        for (int i : nums) {
            freq_table.put(i, freq_table.getOrDefault(i, 0) + 1);
        }

        /* Find the number that only occurs once in Frequency table */
        for (int j : nums) {
            if (freq_table.get(j) == 1) {
                return j;
            }
        }
        return 0;
    }
}